### ISSUE TITLE
Bug of eval loss when enabling average_tokens_across_devices

### ISSUE THREAD

**Issue by Snowdar** [huggingface/transformers#35078]
https://github.com/huggingface/transformers/blob/329f5dbf97a5cb2473914c88c05aa3dcb242e19a/src/transformers/trainer.py#L3737

For the eval set, there is no need to reduce_mean the gradient, so does the loss not need to multiple the `num_processes`?
```
if self.args.average_tokens_across_devices and self.model_accepts_loss_kwargs:
            loss *= self.accelerator.num_processes
```

In my experiments, I found the loss value of eval will be always greater `num_processes` times than the running of `average_tokens_across_devices=false`.

**Comment by Rocketknight1** [comment]
cc @muellerzr @SunMarc

**Comment by techkang** [comment]
I think the bug is caused by the code at https://github.com/huggingface/transformers/blob/v4.46.3/src/transformers/trainer.py#L3661.

Specifically, `num_items_in_batch` is `None` at the eval stage (https://github.com/huggingface/transformers/blob/v4.46.3/src/transformers/trainer.py#L4385) and may be `None` at the training stage (https://github.com/huggingface/transformers/blob/v4.46.3/src/transformers/trainer.py#L5054). `model_accepts_loss_kwargs` is `True` when the model has a `loss_kwargs` argument. `compute_loss_func` is not `None` when users define their own loss function. `num_items_in_batch = self.accelerator.gather(num_items_in_batch).sum().item()` is always triggered at the training stage when `average_tokens_across_devices` is set to `True`.

Therefore, we need to trigger `loss *= self.accelerator.num_processes` only when `num_items_in_batch` is actually used, which is when either `compute_loss_func` is defined or (`model_accepts_loss_kwargs` is `True` and `num_items_in_batch` is not `None`). Changing https://github.com/huggingface/transformers/blob/v4.46.3/src/transformers/trainer.py#L3648-L3651 to something like:
```
if self.args.average_tokens_across_devices and (self.compute_loss_func is not None or (self.model_accepts_loss_kwargs and num_items_in_batch is not None))
```
may temporarily fix this problem. I say "temporarily" because the eval loss will still be different between multi-GPUs and single GPU setups.
